### ISSUE TITLE
Path from onTokenMove should tell if an A* timeout occurred on the path 

### ISSUE THREAD

**Issue by Merudo** [RPTools/maptool#1958]
**Is your feature request related to a problem? Please describe.**
The A* feature is really nice, but a player can exploit it by dropping their token where they shouldn't go.

Even though no path may lead there, once the A* algorithm timeouts the move will be approved and the player will automatically have their token moved there.

There is no easy way for the GM to tell when this occurs, except some intensive and inefficient 
 checks in `onTokenMove`.

**Describe the solution you'd like**
If A* fails to find a path, this fact should be communicated in the `macro.args` of the `onTokenMove` event.

Hence, instead of the `macro.args`  `[{"x":0,"y":300},{"x":250,"y":300}]`, the argument could be

`[{"x":0,"y":300},{"x":250,"y":300, "fail":1}]`

The GM could then deal with this failure as they see fit, perhaps by printing an error message or denying the move outright.

**Comment by Merudo** [comment]
My understanding is that the A* movement occurs through the UI, and then once it is done the code from `onTokenMove` is triggered as a callback.

In other words `onTokenMove` does not move anything, but is simply executed after a move is performed.

Hence I don't think there is a way for `onTokenMove` to specify the timeout.

**Comment by JamzTheMan** [comment]
Not as it's coded currently. It would have to be passed into AStarWalker to override the default timeout value.

I suppose a function/pref could also be used to drive the default timeout but that's a different issue and not sure if its really needed?  

But also at the same time, the failing because of timeout should propagate up to the renderer as well to display something in the UI (another issue is currently opened for this).

I'm thinking I like "status codes" instead of just fail. 404 would be for obvious path not found, but another, more exact reason would be the token is trying to be placed over a cell that is covered by vbl/mbl, so it's an invalid move no matter what vs moving a token to a square surrounded by VBL/MBL.

One may be a "valid" move because of a teleportation and one would be inside a wall.  I think the macro and the UI may want all 3 "status codes" and may be handled differently for each case?

And do we need a generic default for these, whether it's UI or macro? As in the default should be, for players, undo the move and don't allow it vs letting the move happen?

**Comment by Merudo** [comment]
Should there be a "fail" 0/1 together with a "reason" field?

**Comment by Merudo** [comment]
PR #1960 adds a `"fail":1` property to a cell if A* couldn't find  a path to it.

**Comment by JamzTheMan** [comment]
how about "success" or "path found".  It probably should return a value all the time, otherwise you have to check to see if it exists vs just getting the value and testing for a 0 vs 1?

**Comment by Merudo** [comment]
`json.get` always return something even if the property doesn't exist.

`[r: json.get("{}", "fail")]` returns `""`.

**Comment by Phergus** [comment]
This works but seems pretty clunky.

Have to get the last point: `[h: lp = json.get(macro.args, json.length(macro.args) - 1)]`
Then get the fail value:  `[h: f = json.get(lp,"fail)]`
The valid move path success test then looks like: `[r: if(f == "", "success", "fail")]`

Or more ugly:
`[r: if(json.get(json.get(macro.args, json.length(macro.args) - 1),"fail") == "", "success", "fail")]`

As I said it works but seems like the usage will be less than user-friendly for the non-programmers.

**Comment by Merudo** [comment]
Maybe another function should be added? Along with `getLastPath` there could be a function "`isLastPathValid`" that returns a json object containing details on whether the last path was approved by A* or not.

**Comment by Phergus** [comment]
Tested.  Working as described.
